cudamapper/samples/sample_cudamapper.cpp

```diff
@@ -0,0 +1,234 @@
+/*
+* Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+*
+* NVIDIA CORPORATION and its licensors retain all intellectual property
+* and proprietary rights in and to this software, related documentation
+* and any modifications thereto.  Any use, reproduction, disclosure or
+* distribution of this software and related documentation without an express
+* license agreement from NVIDIA CORPORATION is strictly prohibited.
+*/
+
+#include <claraparabricks/genomeworks/cudamapper/cudamapper.hpp>
+#include <claraparabricks/genomeworks/io/fasta_parser.hpp>
+#include <claraparabricks/genomeworks/utils/cudautils.hpp>
+#include <claraparabricks/genomeworks/utils/signed_integer_utils.hpp>
+#include "../src/index_descriptor.hpp"
+#include <cudamapper_file_location.hpp>
+
+#include <claraparabricks/genomeworks/cudamapper/index.hpp>
+#include <claraparabricks/genomeworks/cudamapper/matcher.hpp>
+#include <claraparabricks/genomeworks/cudamapper/overlapper.hpp>
+
+#include <cuda_runtime_api.h>
+#include <iostream>
+#include <string>
+
+using namespace claraparabricks::genomeworks;
+using namespace claraparabricks::genomeworks::cudamapper;
+
+// define constants. See cudamapper/src/application_parameters.hpp
+// constants used in multiple places
+#define INDEX_SIZE 30
+#define KMER_SIZE 15
+#define WINDOWS_SIZE 5
+
+// constants used in the overlapper
+#define MIN_RESIDUES 3
+#define MIN_OVERLAP_LEN 250
+#define MIN_BASES_PER_RESIDUE 1000
+#define MIN_OVERLAP_FRACTION 0.8
+
+// constant used in the indices
+#define FILTERING_PARAMETER 1e-5
+
+// prints overlaps to stdout
+void print_overlaps(const std::vector<Overlap>& overlaps,
+                    const io::FastaParser& query_parser,
+                    const io::FastaParser& target_parser)
+{
+    const int64_t number_of_overlaps_to_print = get_size<int64_t>(overlaps);
+
+    if (number_of_overlaps_to_print <= 0)
+    {
+        std::cout << "There a no overlaps to print!" << std::endl;
+        return;
+    }
+
+    // All overlaps are saved to a single vector of chars and that vector is then printed to output.
+
+    // Allocate approximately 150 characters for each overlap which will be processed,
+    // if more characters are needed buffer will be reallocated.
+    std::vector<char> buffer(150 * number_of_overlaps_to_print);
+    // characters written buffer so far
+    int64_t chars_in_buffer = 0;
+
+    {
```

Comment:
This is a highly optimized version of PAF generating function. As this is supposed to be a simple sample without much caring for the performance I would suggesting going for a simpler solution (even as simple as printing each line individually). An alternative would be to expose cudamapper's function in the public interface, as discussed above

Reply:
My reasoning with using this was so that the output would be in the same format as the main cudamapper binary, and I could sanity check the results easily. I agree that it would be nice to add some sort of formatted print function into the public interface. I think it was mentioned (or included as a bullet point) at the sync yesterday

Reply:
@mimaric I think iterating over the overlaps and doing this: ```
const int32_t added_chars = std::sprintf(buffer.data() + chars_in_buffer,
                                                     "%s\t%lu\t%i\t%i\t%c\t%s\t%lu\t%i\t%i\t%i\t%ld\t%i",
                                                     query_read_name.c_str(),
                                                     query_parser.get_sequence_by_id(overlaps[i].query_read_id_).seq.length(),
                                                     overlaps[i].query_start_position_in_read_,
                                                     overlaps[i].query_end_position_in_read_,
                                                     static_cast<unsigned char>(overlaps[i].relative_strand),
                                                     target_read_name.c_str(),
                                                     target_parser.get_sequence_by_id(overlaps[i].target_read_id_).seq.length(),
                                                     overlaps[i].target_start_position_in_read_,
                                                     overlaps[i].target_end_position_in_read_,
                                                     overlaps[i].num_residues_ * KMER_SIZE, // Print out the number of residue matches multiplied by kmer size to get approximate number of matching bases
                                                     std::max(std::abs(static_cast<int64_t>(overlaps[i].target_start_position_in_read_) - static_cast<int64_t>(overlaps[i].target_end_position_in_read_)),
                                                              std::abs(static_cast<int64_t>(overlaps[i].query_start_position_in_read_) - static_cast<int64_t>(overlaps[i].query_end_position_in_read_))), //Approximate alignment length
                                                     255);
...
printf("%s"..);
```
would be useful in the public API for basic formatted printing

Reply:
I think that the best solution would be to expose `print_paf()` in the public interface (creating a header file for it in `cudamapper/include/claraparabricks/genomeworks/cudamapper/`).
Until this is done I would just add `#include "../src/cudamapper_utils.hpp"` like we do for `IndexDescriptor`

Reply:
Created issue #516

Reply:
Fixed